pkg/controller/resourcetree/tree.go

```diff
@@ -0,0 +1,130 @@
+	ResourceNodeTypeVPC
+	// ResourceNodeTypeCluster represents a EKS cluster resource
+	ResourceNodeTypeCluster
+	// ResourceNodeTypeExternalSecrets represents an External Secrets resource
+	ResourceNodeTypeExternalSecrets
+	// ResourceNodeTypeALBIngress represents an ALB Ingress resource
+	ResourceNodeTypeALBIngress
+	// ResourceNodeTypeExternalDNS represents an External DNS resource
+	ResourceNodeTypeExternalDNS
+	// ResourceNodeTypeGithub represents a Github setup
+	ResourceNodeTypeGithub
+	// ResourceNodeTypeIdentityManager represents a Identity Manager resource
+	ResourceNodeTypeIdentityManager
+	// ResourceNodeTypeArgoCD represents an ArgoCD resource
+	ResourceNodeTypeArgoCD
+)
+
+// ResourceNodeState defines what state the resource is in, used to infer what action to take
+type ResourceNodeState int
+const (
+	// ResourceNodeStateNoop represents a state where no action is needed. E.g.: if the desired state of the
+	// resource conforms with the actual state
+	ResourceNodeStateNoop ResourceNodeState = iota
+	// ResourceNodeStatePresent represents the state where the resource exists
+	ResourceNodeStatePresent
+	// ResourceNodeStateAbsent represents the state where the resource does not exist
+	ResourceNodeStateAbsent
+)
+
+// CommonMetadata represents metadata required by most if not all operations on services
+type CommonMetadata struct {
+	Ctx context.Context
+	Id api.ID
+}
+
+// StateRefreshFn is a function that attempts to retrieve state potentially can only be retrieved at runtime. E.g.:
+// state that can only exist after an external resource has been created
+type StateRefreshFn func(node *ResourceNode)
+
+// ResourceNode represents a component of the cluster and its dependencies
+type ResourceNode struct {
+	Type  ResourceNodeType
+	State ResourceNodeState
+
+	// Contains metadata regarding the resource supplied by the desired state definition
+	Metadata             interface{}
+
+	StateRefresher 		 StateRefreshFn
+	// ResourceState contains data that needs to be retrieved runtime. In other words, data that potentially only exist
+	// after an external resource has been created
+	ResourceState 		 interface{}
+
+	Children []*ResourceNode
+}
+
+// RefreshState calls the stored StateRefreshFn if it exists
+func (node *ResourceNode) RefreshState() {
+	if node.StateRefresher == nil {
+		return
+	}
+
+	node.StateRefresher(node)
+}
+
+// SetStateRefresher stores a StateRefreshFn on the node to be used to retrieve runtime state later
+func (node *ResourceNode) SetStateRefresher(nodeType ResourceNodeType, refresher StateRefreshFn) {
+	targetNode := node.GetNode(&ResourceNode{Type: nodeType})
+
+	if targetNode == nil {
+		return
+	}
+
+	targetNode.StateRefresher = refresher
+}
+
+// Equals knows how to compare two ResourceNodes and determine equality
+func (node *ResourceNode) Equals(targetNode *ResourceNode) bool {
+	if node == nil {
+		return false
+	}
+
+	// P.S.: For now, this is good enough due to all types of resources only existing once
```

Comment:
For now that is just future proofing. However, I've had it in mind (therefore the comment) and I don't think it would be that hard to implement.

Something like this in func (node *ResourceNode) Equals(..)
```golang
if node.equalityFn != nil {
    return node.equalityFn(targetNode)
}

return node.Type == targetNode.Type
```

and set the equalityFn when we create the node / during applyMetadata

Reply:
As discussed on Discord, I was more interested in the probability that this happens. The answer here is that cluster.yaml (the desired state) doesn't currently allow defining multiple resources, as it is represented as yaml maps.